Deploy

VictorTikhonov opened the pull request:
Добавлены DTO для аналитики расходов
1.1. TravelExpenseAnalyticsDTO — ответ по поездке: общая сумма расходов + список категорий
1.2. CategoryAnalyticsResponseDTO — детализация по каждой категории: id и название категории, общая сумма, процент от общей суммы поездки, количество трат, статистика по участникам (кто сколько потратил)
1.3. Внутренний класс ParticipantStats — имя участника, фамилия и его сумма расходов внутри категории

Создан новый REST-контроллер AnalyticsController
2.1. Эндпоинт GET /api/travels/{travelId}/analytic возвращает аналитику расходов по указанной поездке
2.2. Добавлена авторизация: доступ только для участников поездки (@PreAuthorize("@travelSecurity.isMember(..."))

Реализован сервис ExpenseAnalyticsService
3.1. Проверяется существование поездки
3.2. Загружаются все категории и все расходы по поездке
3.3. Расходы группируются по категориям (категория может быть null)
3.4. Рассчитывается: сумма расходов в каждой категории, общий расход по поездке, процент каждой категории от всей потраченной суммы, количество трат в категории, траты каждого участника в рамках категории
3.5. Добавлена логика расчёта статистики участников: для каждой категории собираются доли участников из MemberExpense, в статистику включаются только те участники, чьи траты > 0
3.6. Подготовлены удобные вспомогательные методы: группировка расходов по категориям, подсчёт общей суммы расходов в категории, подсчёт количества трат, генерация DTO для ответа, сбор статистики по пользователям

Добавлены DTO для создания, обновления и возврата расходов
4.1. ExpenseRequestDTO используется при создании расхода и содержит название, описание, дату (по умолчанию текущее время), категорию, плательщика, participantShares — доли всех участников. Добавлена валидация: обязательные поля, непустые доли, положительные значения
4.2. ExpenseUpdateRequestDTO — DTO для частичного обновления расхода, все поля опциональны
4.3. ExpenseResponseDTO и MemberExpenseResponseDTO — DTO для выдачи клиенту, MemberExpenseResponseDTO имеет корректный equals/hashCode на основе userId

Создан контроллер ExpenseController
5.1. Реализованы следующие эндпоинты (все защищены @PreAuthorize — доступ только участникам поездки):
5.1.1. POST /expenses — создание расхода
5.1.2. PATCH /expenses/{expenseId} — частичное обновление
5.1.3. DELETE /expenses/{expenseId} — удаление расхода
5.1.4. DELETE /expenses/{expenseId}/participants — удаление участников
5.1.5. POST /expenses/{expenseId}/participants — добавление участников
5.2. Все методы возвращают актуальный ExpenseResponseDTO

Основная бизнес-логика реализована в ExpenseService
6.1. Создание расхода
6.1.1. Проверяется существование поездки и категории, что все участники являются членами поездки, что плательщик входит в участников, что все доли положительны
6.1.2. Сумма рассчитывается автоматически как сумма долей
6.1.3. Создаётся Expense и распределяются доли: плательщик — положительная доля, остальные — отрицательная (долг)
6.1.4. Формируется корректный DTO
6.2. Обновление расхода
6.2.1. Поддерживаются частичные обновления: основные поля (name, description, category, date), плательщик (смена возможна, пересчитываются знаки долей), доли участников (только уже существующих — новые добавляются отдельным endpoint)
6.2.2. Проверки: DTO не пустой (если нет ни одного поля — ошибка), доли валидируются, нельзя обновлять долю несуществующего участника
6.2.3. Сумма пересчитывается как сумма модулей всех долей
6.3. Удаление расхода
6.3.1. Удаление расхода из поездки, проверяется, что указанный расход принадлежит данной поездке
6.4. Удаление участников из расхода
6.4.1. Проверяется: список не пуст, все участники действительно участвуют в расходе, плательщика удалить нельзя, после удаления пересчитывается сумма
6.5. Добавление новых участников
6.5.1. Проверяется: список не пуст, все новые участники входят в поездку, ни один не дублирует уже существующих участников, доли валидны
6.5.2. Добавляемые доли автоматически становятся отрицательными (это долг участника), сумма соответственно увеличивается

Помощники и валидация
7.1. validateExpenseRequestDTO() — комплексная валидация при создании
7.2. validateShares() — проверка корректности долей
7.3. validateParticipantSharesForUpdate() — проверка при обновлении
7.4. updateBasicInfo() — безопасное обновление полей
7.5. updateFinancialFields() — пересчёт долей и суммы
7.6. updatePayer() — логика смены плательщика
7.7. ensureParticipantsAreNew() — защита от дубликатов
7.8. findExpenseInTravel() — гарантирует существование расхода в поездке

Добавлен сервис MemberExpenseService
8.1. Метод addMemberExpense добавляет участника траты
8.2. Метод removeMemberExpense удаляет участника траты

Добавлен сервис UserService
9.1. Метод getUsers возвращает список пользователей по множеству идентификаторов

Добавлен сервис CategoryService
10.1. Метод findAllCategoryInTravel возвращает все категории в поездке
10.2. Метод findCategory ищет категорию по id и выбрасывает CategoryNotFoundException при отсутствии

Добавлен компонент ExpenseDtoMapper
11.1. Метод createExpenseResponseDTO формирует ExpenseResponseDTO из сущности Expense
11.2. Метод createSetParticipantsResponseDTO создаёт Set<MemberExpenseResponseDTO> из набора MemberExpense
11.3. Метод createMemberExpenseResponseDTO формирует DTO участника с id, именем, фамилией и положительной долей

Добавлен глобальный обработчик исключений RestExceptionHandler

TravelService
13.1. Создание/редактирование/закрытие/удаление поездок, проверка и валидация дат
13.2. Методы для добавления/удаления участников, трат и переводов с привязкой к поездке
13.3. Поиск и сохранение поездок

TravelMemberService
14.1. Приглашение участников по телефону, ответ на приглашения, получение приглашений/списка участников
14.2. Кик/выход с защитой владельца, проверки наличия пользователя в поездке
14.3. Получение поездок пользователя

TravelController (/api/travels)
15.1. Создание поездки
15.2. Получение списка моих поездок
15.3. Детальная поездка
15.4. Редактирование, закрытие/переоткрытие и удаление

TravelMemberController (/api/members)
16.1. Приглашение пользователей по телефонам в поездку
16.2. Получение приглашений
16.3. Ответ на приглашения
16.4. Просмотр участников поездки
16.5. Кик участника, выход участника

Добавлены фабрики для создания объектов

EnzoLu1337 committed:
Init Database

EnzoLu1337 committed:
Add entity and repository for User and PasswordCredential

EnzoLu1337 committed:
Add enum, entity and repository for Travel

EnzoLu1337 committed:
Add enum, entity and repository for TravelMember

EnzoLu1337 committed:
Clean DDL details

VictorTikhonov committed:
Редактирование файла V1_init.sql

- expens заменено на expense
- recieve_id заменено на recipient_id

VictorTikhonov committed:
Создал сущность "Трата" и соответсвующий репозиторий

VictorTikhonov committed:
Создал сущность "История", соответсвующий репозиторий и тип перечисления

VictorTikhonov committed:
Создал сущность "Участники траты" и соответсвующий репозиторий

EnzoLu1337 committed:
Исправление работы Entity с ENUM

QwantXG committed:
Add Category, Event, Transfer entities and repositories

VictorTikhonov committed:
Измение класса Expense

- Добавил поле "Category"
- Изменил фабричный метод формирования траты, добавим параметр категория

VictorTikhonov committed:
Рефакторинг сущностей Expense, MemberExpense и History:

- В MemberExpense убран публичный сеттер expense, теперь поле изменяется только через addMemberExpense/removeMemberExpense в Expense
- Фабричные методы create для Expense и MemberExpense используют сеттеры с проверкой
- Сеттеры sum, payer, travel, share, author, type созданы с проверкой, чтобы защищать состояние объектов
- Убраны лишние Lombok @Setter и @Setter(AccessLevel.NONE)

EnzoLu1337 committed:
Исправление названия столбца с recieve_id на recipient_id

EnzoLu1337 committed:
Добавлены DTO для авторизации и регистрации

EnzoLu1337 committed:
Добавлены настройки и сервис для JWT

EnzoLu1337 committed:
Убраны лишние импорты из JwtService

EnzoLu1337 committed:
Добавлена аннотация для конфигурации JWT, добавлены миграция, сущность и репозиторий для refresh токена

EnzoLu1337 committed:
Перенос JwtService в папку со всем security

EnzoLu1337 committed:
Добавлен фильтр для аутентификации JWT

EnzoLu1337 committed:
Добавлено поле MemberRole в TravelMember

EnzoLu1337 committed:
Добавлено enum MemberRole

EnzoLu1337 committed:
Добавлен UserDetailsServiceImpl и UserPrincipal для аутентификации пользователей по номеру телефона

EnzoLu1337 committed:
Добавлен класс SecurityConfig для настройки безопасности и аутентификации с использованием JWT

EnzoLu1337 committed:
Добавлен контроллер аутентификации и сервис для регистрации и входа пользователей с использованием JWT

EnzoLu1337 committed:
Добавлены DTO классы для изменения пароля и обновления/выхода из сессии

EnzoLu1337 committed:
Удалены методы для парсинга refresh токена и проверки истечения access токена в JwtService

EnzoLu1337 committed:
Добавлены классы AccountController и AccountService для управления изменением пароля, выходом из системы и обновлением токенов

EnzoLu1337 committed:
Добавлен обработчик исключений для управления ошибками аутентификации и валидации

EnzoLu1337 committed:
Упрощены сообщения об ошибках валидации для запросов аутентификации и регистрации

EnzoLu1337 committed:
Добавлены зависимости для Bouncy Castle и обновлены параметры кодирования паролей, а также добавлены временные метки для токенов

EnzoLu1337 committed:
Добавлены обработчики исключений для некорректного JSON, неподдерживаемых методов и отсутствующих эндпоинтов; разрешен доступ к маршруту /error

EnzoLu1337 committed:
Рефакторинг обработки исключений: перемещены классы исключений в общий пакет; добавлены новые исключения для пользователей и путешествий

EnzoLu1337 committed:
Добавлены DTO для создания и получения путешествий, а также валидатор для проверки диапазона дат

EnzoLu1337 committed:
Изменен маршрут для получения путешествий на /mytravels; добавлен статус ACTIVE при создании путешествия; добавлены аннотации @Transactional для методов получения путешествий и участников

EnzoLu1337 committed:
Добавлены DTO для редактирования путешествий и обработки приглашений; реализованы методы для редактирования, закрытия, повторного открытия и удаления путешествий; добавлены исключения для некорректного диапазона дат и недопустимого удаления владельца; обновлены обработчики безопасности для проверки статуса участников.

EnzoLu1337 committed:
Переименованы классы исключений UserNotFound и TravelNotFound в UserNotFoundException и TravelNotFoundException

EnzoLu1337 committed:
Добавлено новое исключение ConflictStateException для обработки конфликтов состояния поездки

EnzoLu1337 committed:
Добавлено получение информации о текущем пользователе

EnzoLu1337 committed:
Добавлено выбрасывание исключения при попытке повторного приглашения пользователя в поездку

EnzoLu1337 committed:
Изменено приглашение участников: теперь поддерживается список номеров телефонов

EnzoLu1337 committed:
Изменено сообщение об ошибке для некорректного номера телефона в InviteRequest

VictorTikhonov committed:
Реализация управления расходами: добаление и удаление траты

ExpenseController:
- POST /travels/{travelId}/expenses: создает расход для поездки
- DELETE /travels/{travelId}/expenses/{expenseId}: удаляет расход в поездке

ExpenseService:
- createExpense:
    - создает расход
    - распределяет доли участников: каждому участнику создается MemberExpense с отрицательной суммой, плательщику добавляется остаток
    - сохраняет расход
    - возвращает ExpenseResponseDTO с участниками
- distributeExpenseShares: приватный метод для вычисления долей участников и остатка плательщика
- checkShare: проверяет корректность доли участника
- createExpenseResponseDTO + createSetParticipantsResponseDTO + createMemberExpenseResponseDTO: формируют DTO для API
- deleteExpense: удаляет расход после проверки принадлежности к поездке
- findUserInTravel, findExpenseInTravel, findTravel, findUser, findCategory: вспомогательные методы для проверки существования сущностей

DTO:
- ExpenseRequestDTO: данные для создания расхода, с валидацией полей
- ExpenseResponseDTO (record): данные о расходе, включая участников
- MemberExpenseResponseDTO (record): данные об участниках, equals/hashCode по userId

Exceptions:
- ExpenseNotInTravelException
- InvalidParticipantShareException
- PayerShareInvalidException
- UserNotInTravelException
- TravelNotFoundException
- UserNotFoundException
- CategoryNotFoundException

- Repository (добавлены методы):
  - ExpenseRepository: findByIdAndTravelId— проверяет принадлежность расхода к поездке
  - TravelMemberRepository: findByUserIdAndTravelId — проверяет, что пользователь участвует в поездке

QwantXG committed:
Add transfer

QwantXG committed:
корректное подк exception

QwantXG committed:
Проверка наличия поездки

VictorTikhonov committed:
Добавлен сервис для проверки сущностей и прав доступа

- Проверка пользователей в поездках
- Валидация существования поездок и категорий
- Централизованная обработка исключений

VictorTikhonov committed:
Сообщения в исключениях переведены на русский

VictorTikhonov committed:
Переработана обработка ошибок, документация и DTO обновления расходов

- Переименованы исключения:
  - ExpenseNotInTravelException на ExpenseNotFoundInTravelException
  - UserNotInTravelException на UserNotFoundInTravelException
- Добавлены JavaDoc-комментарии ко всем методам ExpenseService
- Уточнена логика обновления расходов:
  - добавлена проверка пустого DTO (EmptyUpdateRequestException)
  - уточнена проверка обновления финансовых полей (sum и participantShares)
- Создан ExpenseUpdateRequestDTO: разделены «безопасные» и «финансовые» поля
- Уточнены сообщения исключений и внутренняя логика валидации

VictorTikhonov committed:
Добавлен контроллер для обработки исключений

VictorTikhonov committed:
Удалена передача суммы в ExpenseRequestDTO и сервис createExpense

Теперь сумма расхода вычисляется автоматически как сумма долей всех участников.

VictorTikhonov committed:
Добавлена возможность явно указывать плательщика при создании траты

- В DTO ExpenseRequestDTO добавлено поле payerId (обязательное)
- При создании траты:
  * Плательщик получает положительную долю
  * Остальные участники получают отрицательные доли
  * Сумма траты вычисляется как сумма всех долей участников (sum больше не передается клиентом)
- Валидируются:
  * Все участники находятся в поездке
  * Плательщик присутствует среди участников
  * Все доли положительные
- Добавлены новые исключения:
  * InitiatorNoAccessException — если инициатор не имеет доступа к поездке
  * PayerNotInParticipantsException — если плательщик не указан среди участников
- Методы ReferenceLookupService расширены для проверки доступа инициатора и валидности участников
- Рефакторинг методов распределения долей и вычисления суммы для ясности и предсказуемости

VictorTikhonov committed:
Обновление логики редактирования траты и DTO

- В ExpenseResponseDTO добавлено поле payerId для возврата идентификатора плательщика.
- ExpenseUpdateRequestDTO изменён:
  * удалено поле sum (теперь сумма вычисляется автоматически)
  * добавлено поле payerId для возможности смены плательщика при редактировании
- Логика updateExpense:
  * теперь можно менять базовую информацию, доли существующих участников и плательщика
  * доли участников обновляются только для существующих участников; новые добавить нельзя
  * при смене плательщика пересчитываются знаки долей: новый плательщик положительный, остальные — отрицательные
  * пересчитывается сумма траты как сумма модулей всех долей
  * валидация participantShares учитывает только существующих участников
- Добавлены и улучшены вспомогательные методы:
  * updatePayer, updateFinancialFields, validateParticipantSharesForUpdate, checkUserInExpense, hasAnyField
- Исключено использование checkInitiatorAccessToTravel из ReferenceLookupService (лишнее)
- Документация методов обновлена, описывает поведение при редактировании траты и правила валидации

VictorTikhonov committed:
Новое искючение если пользователь не найден в трате

Забыл добавить в прошлый коммит!

VictorTikhonov committed:
Добавлен функционал удаления участников из траты

- Реализован метод deleteParticipantsFromExpense в ExpenseService
- Проверка наличия участников в тратах (validateUsersInExpense)
- Защита от удаления плательщика
- Перерасчёт суммы траты после удаления участников
- Добавлен эндпоинт DELETE /api/travels/{travelId}/expenses/{expenseId}/participants
- Добавлен метод getMemberExpense в ReferenceLookupService для поиска участника траты
- Удалены устаревшие методы checkTravelExists и checkUserExists из ReferenceLookupService
- Добавлен метод findByParticipantIdAndExpenseId в MemberExpenseRepository
- Исключения обработаны через глобальный ExceptionHandler

VictorTikhonov committed:
Добавлен функционал добавления участников к существующей трате

- Реализован метод addParticipantsToExpense в ExpenseService
- Проверка, что все переданные пользователи являются участниками поездки
- Проверка на дублирующихся участников через ensureParticipantsAreNew
- Доли участников должны быть валидными (не null, > 0)
- Создание новых MemberExpense
- Пересчет суммы траты после добавления новых участников
- Добавлен POST эндпоинт /api/travels/{travelId}/expenses/{expenseId}/participants
- Добавлен метод получения пользователей по Id в ReferenceLookupService

VictorTikhonov committed:
Удалени пометки не актуальные TODO

EnzoLu1337 committed:
Рефакторинг TravelService: добавлены TravelFactory и TravelMapper для упрощения создания и обработки поездок

EnzoLu1337 committed:
Изменены HTTP методы в TravelController

EnzoLu1337 committed:
Merge pull request #27 from tbank-ttravels/travels

Travels

EnzoLu1337 committed:
Merge branch 'deploy' into expenses-management

EnzoLu1337 committed:
Merge pull request #28 from tbank-ttravels/expenses-management

Expenses management

EnzoLu1337 committed:
Merge branch 'deploy' into transfers

EnzoLu1337 committed:
Merge pull request #29 from tbank-ttravels/transfers

Transfers

EnzoLu1337 committed:
Рефакторинг хендлера исключений, исправление использования исключений

EnzoLu1337 committed:
Изменен путь API в TravelController и перемещена логика обновления поездок в TravelService

EnzoLu1337 committed:
Удален тип токена по умолчанию из AuthResponse

EnzoLu1337 committed:
Фабрика Travel переведена в static

EnzoLu1337 committed:
Добавлен обработчик исключений для валидации

EnzoLu1337 committed:
Добавлен файл README.md

EnzoLu1337 committed:
Отделен сервис для работы с участниками траты, так же вынесена в функции работа с репозиториями.

EnzoLu1337 committed:
Добавлен зависимость springdoc-openapi для генерации документации API

EnzoLu1337 committed:
Добавлено временное разрешение доступа к Swagger UI для тестирования

EnzoLu1337 committed:
Добавлен класс ErrorResponse для обработки ошибок и конфигурация OpenAPI для документации API

EnzoLu1337 committed:
Добавлено аннотирование @Transactional для метода getMyTravels и удалено из методов findAllTravelForUser и findUserInTravel

EnzoLu1337 committed:
Добавлено аннотирование @Transactional для метода getMyTravels и удалено из методов findAllTravelForUser и findUserInTravel

EnzoLu1337 committed:
Добавлено документирование API для контроллера Travel

EnzoLu1337 committed:
Добавлено документирование API для контроллера Travel

EnzoLu1337 committed:
Добавлено документирование API для контроллера TravelMember

EnzoLu1337 committed:
Добавлено документирование API для контроллера TravelMember

EnzoLu1337 committed:
Добавлены недостающие описания для pathVariable

EnzoLu1337 committed:
Добавлены недостающие описания для pathVariable

EnzoLu1337 committed:
Переведены ответы exception на DTO

EnzoLu1337 committed:
Переведены ответы exception на DTO

EnzoLu1337 committed:
Добавлено документирование API для контроллеров Auth и Account
Разделены на два разных DTO LogoutRequest и RefreshRequest
Косметические изменения в TravelController и TravelMemberController

EnzoLu1337 committed:
Добавлено документирование API для контроллеров Auth и Account
Разделены на два разных DTO LogoutRequest и RefreshRequest
Косметические изменения в TravelController и TravelMemberController

VictorTikhonov committed:
Исправил, согласно комментам

- Добавил фабрики HistoryFactory, ExpenseFactory, MemberExpenseFactory
- Добавлены аннотации builder и setter над классами History, Expense, MemberExpense
- Соответственно изменен сервис

VictorTikhonov committed:
Добавлена аналитика расходов по категориям

1. Сервис ExpenseAnalyticsService:
- Реализован метод analyticsCategory() для формирования аналитики расходов по категориям в рамках одной поездки.
- Рассчитывается общая сумма, доли категорий, количество трат и статистика по участникам.
- Добавлены вспомогательные методы: группировка трат по категориям, подсчёт сумм, формирование статистики участников.
2. ExpenseRepository и ExpenseService:
- В репозитории добавлен метод findAllByTravelId(Long travelId).
- В сервисе добавлен метод findAllExpenseInTravel(Long travelId) для вызова репозитория.
3. ReferenceLookupService:
- Добавлен метод findAllCategoryInTravel(Long travelId) для получения категорий поездки.
- Добавлен метод findAllUsersInTravel(Long travelId) для получения участников поездки.
- Добавлен метод checkTravel(Long travelId) для проверки существования поездки.
4. DTO:
- Создан TravelExpenseAnalyticsDTO с общей суммой расходов и списком категорий.
- Дополнен CategoryAnalyticsResponseDTO для включения участников и сумм по категории.
5. Контроллер:
- Создан AnalyticsController с endpoint GET /api/travels/{travelId}/analytic.
6. Документация:
- Написан подробный JavaDoc ко всем методам сервиса.

VictorTikhonov committed:
Редактирование

- Теперь некоторые исключения принимают сообщение вместо id
- А контроллере добавлен PreAuthorize
- В фабрике убран неиспользующийся код
- ExpenseRequestDTO отредактированы сообщения об ошибках в аннотациях
- ExpenseService теперь использует TravelMemberService и TravelService
- Удален конструктор, добавлена аннотация RequiredArgsConstructor
- В Travel добавлен список трат
- В TravelService добавлены методы добавления/удаления траты

VictorTikhonov committed:
Обновление работы с исключениями

- Теперь исклюения принимают String
- Соответствено изменено их создание при выбрасывнии
- Удалены неиспользующиеся исключения
- RestExceptionHandler соответственно обновлен
- ExpenseAnalyticsService теперь зависит от TravelService и TravelMemberService
- В ReferenceLookupService убраны не использующиеся методы, доабвлена аннотация RequiredArgsConstructor, удален конструктор

VictorTikhonov committed:
Merge pull request #32 from tbank-ttravels/expenses

Expenses

VictorTikhonov committed:
Удалены не нужные методы в репозиториях

VictorTikhonov committed:
RestExceptionHandler перенесен в пакет с контроллерами. Удалены не нужные строки

EnzoLu1337 committed:
Добавил в TravelService методы для добавления/удаления участников через Travel и обновил TravelMemberService, чтобы приглашения/удаления сохранялись через поездку, а не напрямую в репозиторий.

EnzoLu1337 committed:
Добавил поддержку переводов через агрегат поездки — коллекция transfers в Travel, методы add/removeTransfer в TravelService, а TransferService теперь добавляет/сохраняет переводы через Travel и проверяет участников поездки.

VictorTikhonov committed:
Добавлены сервисы UserService и CategoryService, удален ReferenceLookupService.

- Теперь в контрлерах вместо ReferenceLookupService, используются UserService и CategoryService

VictorTikhonov committed:
Изменена опечатка в возвращаемом ответе

QwantXG committed:
add categoryservice

EnzoLu1337 committed:
Merge remote-tracking branch 'origin/swagger-test-travel' into swagger-test-travel

# Conflicts:
#	src/main/java/com/tbank/ttravels_backend/controller/RestExceptionHandler.java

EnzoLu1337 committed:
Добавлены модульные тесты TravelService
create/get/edit/close/reopen/delete
add/remove expense/transfer/member
валидация дат и дубликатов
проверки not found/конфликтов

EnzoLu1337 committed:
Добавлены модульные тесты TravelMemberService

EnzoLu1337 committed:
Добавлены модульные тесты AccountService

nbiryulin commented on README.md:
```diff
@@ -0,0 +1,34 @@
+
+   spring.datasource.url=jdbc:postgresql://localhost:5432/ttravels_db
+   spring.datasource.username=postgres
+   spring.datasource.password=postgres
```
просто fyi пароли так никогда не стоит хранить, самый простой способ - в локальный файл, добавленный  .gitignore . Ну и в readme тем более не стоит добавлять)

nbiryulin commented on src/main/java/com/tbank/ttravels_backend/controller/AnalyticsController.java:
```diff
@@ -0,0 +1,29 @@
+    public TravelExpenseAnalyticsDTO analytic(@PathVariable Long travelId,
+                                              @AuthenticationPrincipal UserPrincipal principal) {
+
+        return this.expenseAnalyticsService.analyticsCategory(travelId);
```
1. зачем на 27 строке this. 
2. метод analyticsCategory давайте переименуем. Из названия непонятно, что он делает. Должно быть какое-то действие в названии 
3. Необязательно: в идеале выносить общее начало API куда-нибудь в общую константу. Так как иначе, если что-то поменяется , а проект будет большим - замучаетесь менять

nbiryulin commented on src/main/java/com/tbank/ttravels_backend/controller/ExpenseController.java:
```diff
@@ -0,0 +1,86 @@
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    @PreAuthorize("@travelSecurity.isMember(#travelId, principal.id)")
```
а для создателя поездки  isMember же тоже true будет?

nbiryulin commented on src/main/java/com/tbank/ttravels_backend/controller/TransferController.java:
```diff
@@ -0,0 +1,36 @@
+        return transferService.editTransfer(id, request);
+    }
+
+    @GetMapping("/{travelId}/transfers")
```
/transfers два раза в API дублируется, будто бы излишне

nbiryulin commented on src/main/java/com/tbank/ttravels_backend/controller/TravelController.java:
```diff
@@ -0,0 +1,77 @@
+    private final TravelService travelService;
+    private final TravelMemberService travelMemberService;
+
+    @PostMapping("/create")
```
давайте приведем код к одному виду 
1) Необязательно: ВЕЗДЕ использовать @RequiredArgsConstructor или самостоятельно задавать конструктор 
2) Обазятельно: API к одному виду привести, иначе сами себе в ногу выстрелите. Либо везде добавлять /create и прочие действия, либо везде одинаковые URI , различающиеся http методами

nbiryulin commented on src/main/java/com/tbank/ttravels_backend/controller/TravelController.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @PatchMapping("/{travelId}")
+    @PreAuthorize("@travelSecurity.isOwner(#travelId, principal.id)")
```
только owner может менять travel?

nbiryulin commented on src/main/java/com/tbank/ttravels_backend/controller/TravelMemberController.java:
```diff
@@ -0,0 +1,69 @@
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/api/members")
```
как будто бы стоит добавить travel в API, так как к ней относится. Причем, если используете {travelId} оно должно идти как /travel/{travelId}/member, например

nbiryulin commented on src/main/java/com/tbank/ttravels_backend/security/AccountController.java:
```diff
@@ -0,0 +1,43 @@
+package com.tbank.ttravels_backend.security;
```
необязательно: лучше вынести controller-ы и сервисы отсюда в controller и service , а уже там можно создать security. в случае большого проекта так гораздо удобнее искать

nbiryulin commented on src/main/java/com/tbank/ttravels_backend/controller/AccountController.java:
```diff
@@ -0,0 +1,43 @@
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/account")
```
ну и API к одному ввиду - во всех остальных стартует с API

EnzoLu1337 committed:
Исправление ошибок
Вынесена общее начало API в общую константу
API приведено к общему виду

EnzoLu1337 committed:
Merge branch 'deploy' into swagger-test-travel

# Conflicts:
#	src/main/java/com/tbank/ttravels_backend/controller/AccountController.java
#	src/main/java/com/tbank/ttravels_backend/controller/AuthController.java
#	src/main/java/com/tbank/ttravels_backend/controller/TravelController.java
#	src/main/java/com/tbank/ttravels_backend/controller/TravelMemberController.java
#	src/main/java/com/tbank/ttravels_backend/service/TravelMemberService.java

VictorTikhonov committed:
Добавлен swagger на ExpenseController и AnalyticExpenseController. Также на соответствующие ДТО

- Переименовал AnalyticsController на AnalyticExpenseController

VictorTikhonov committed:
Добавлен swagger на ExpenseController и AnalyticExpenseController. Также на соответствующие ДТО

- Переименовал AnalyticsController на AnalyticExpenseController

VictorTikhonov committed:
Добавлен эндпоинт для возврата все трат в поездке

- Создано новое ДТО TravelExpensesResponseDTO
- Создан эндпоинт для получения всех трат
- Проставленны аннотации Swagger
- Исправлены недочеты (забыл проставить анноатиции свагера над ДТО некоторыми, изменен метод репозитория...)

VictorTikhonov committed:
Добавлен эндпоинт для возврата все трат в поездке

- Создано новое ДТО TravelExpensesResponseDTO
- Создан эндпоинт для получения всех трат
- Проставленны аннотации Swagger
- Исправлены недочеты (забыл проставить анноатиции свагера над ДТО некоторыми, изменен метод репозитория...)

EnzoLu1337 committed:
Рефакторинг

EnzoLu1337 committed:
Рефакторинг

EnzoLu1337 committed:
Merge remote-tracking branch 'origin/expenses_swagger' into expenses_swagger

# Conflicts:
#	src/main/java/com/tbank/ttravels_backend/controller/ExpenseController.java

EnzoLu1337 committed:
Merge pull request #36 from tbank-ttravels/expenses_swagger

Expenses swagger

EnzoLu1337 committed:
Добавлено общее начало api в глобальный конфиг

EnzoLu1337 committed:
Merge branch 'swagger-test-travel' into deploy

EnzoLu1337 committed:
Исправление после слияния

EnzoLu1337 committed:
Исправление тестов после изменения удаления пользователя из поездки

EnzoLu1337 committed:
Удалены тесты

QwantXG committed:
Update V3__member_role.sql

QwantXG committed:
Добавлены сервисы и контроллеры

EnzoLu1337 committed:
Merge branch 'deploy' into transfers

# Conflicts:
#	src/main/java/com/tbank/ttravels_backend/controller/RestExceptionHandler.java
#	src/main/java/com/tbank/ttravels_backend/service/CategoryService.java

EnzoLu1337 committed:
Возврат к состоянию ветки Deploy

EnzoLu1337 committed:
Изменения для deploy
На дто повешаны аннотации
Контроллер приведён к общему виду

EnzoLu1337 committed:
Изменения для deploy
На дто повешаны аннотации
Контроллер приведён к общему виду

EnzoLu1337 committed:
Фикс после изменений.
Убрана проверка travelId из request

EnzoLu1337 committed:
Убран лишний /api

EnzoLu1337 committed:
Исправления ошибки доступа

EnzoLu1337 committed:
Добавлено новое условие для проверки статуса участника поездки

EnzoLu1337 committed:
Merge pull request #37 from tbank-ttravels/transfers

Transfers

EnzoLu1337 committed:
Добавлено окружение для Docker и обновлена документация

VictorTikhonov committed:
Реализован полный расчёт долгов и возврат структуры долгов по поездке

- Добавлена сервисная логика DebtCalculationService:
  • расчёт задолженностей по расходам пользователя;
  • учёт переводов (входящих/исходящих);
  • учёт расходов, где пользователь является должником;
  • нормализация балансов и формирование итогового ответа.

- Добавлены DTO-модели:
  • UserDTO — представление пользователя;
  • DebtInfo — информация о должнике/кредиторе;
  • TravelDebtsResponse — итоговый ответ с двумя списками (должники/кредиторы).

- Добавлен UserDtoMapper для преобразования сущностей User в DTO.

- DebtController: возвращает рассчитанные долги текущего пользователя.

- Реализованы необходимые методы в ExpenseService, TransferService и репозиториях.

- У сущности User добавлены корректные equals()/hashCode() по id.

VictorTikhonov committed:
Добавлен swagger и другое

- теперь дто стали record
- изменено название некоторых дто
- добавлен swagger для контроллера и дто

VictorTikhonov committed:
Merge pull request #38 from tbank-ttravels/Debt_Calculation

Debt calculation

VictorTikhonov committed:
Добавлена документация и параметры поменялись местами в методе findExpensesWhereUserIsDebtor

VictorTikhonov committed:
В Travel добавлена аннотация @Builder.Default и @Builder в ExpenseUpdateRequestDTO

VictorTikhonov committed:
Тесты для создания траты

VictorTikhonov committed:
Тесты для добавления новых участников в существующую трату

VictorTikhonov committed:
Тесты для удаления траты

VictorTikhonov committed:
Тесты для удаления участников из траты

VictorTikhonov committed:
Тесты для обновления траты

VictorTikhonov committed:
Фабрика создания объектов для тестов, связанных с тратами

VictorTikhonov committed:
Тесты для получения всех трат

VictorTikhonov committed:
Тесты для подсчета долгов пользователя

EnzoLu1337 committed:
Merge branch 'docker' into deploy

EnzoLu1337 committed:
Добавлено новое исключение для обработки ошибки UserNotFoundByPhoneException

EnzoLu1337 committed:
Добавлена возможность заново пригласить пользователя

EnzoLu1337 committed:
В список поездок пользователя добавлена его роль

EnzoLu1337 committed:
Роль пользователя перенесена с получения поездок в получение участников поездки

VictorTikhonov committed:
Тесты, связанные с тратами, перенесены в отдельный пакет

EnzoLu1337 committed:
Добавлена возможность изменения информации о пользователе и обновлен метод получения текущего пользователя

EnzoLu1337 committed:
Исправлена проблема с Lazy связью при запросе трат

EnzoLu1337 committed:
Исправлена проблема что в ответе был id null

EnzoLu1337 committed:
Добавлена проверка на состояние поездки

VictorTikhonov committed:
Тесты для работы с категориям трат

VictorTikhonov committed:
Тесты для работы с переводами

VictorTikhonov committed:
Тесты для работы с участникми траты

VictorTikhonov committed:
Добавлен тест для при создании траты

VictorTikhonov committed:
Тесты для аналитики

VictorTikhonov committed:
Тесты для работы с участниками поездки + изменена фабрика

VictorTikhonov committed:
Тесты для работы с поездкой + изменена фабрика

VictorTikhonov committed:
Тесты безопасности для работы с JwtService

VictorTikhonov committed:
Тесты безопасности для работы с TokenHashService

VictorTikhonov committed:
Тесты безопасности для работы с UserDetailsServiceImpl + изменена фабрика для тестов

VictorTikhonov committed:
Тесты безопасности для работы с TokenIssuer + изменена фабрика для тестов

VictorTikhonov committed:
Правильные тесты для поездки

VictorTikhonov committed:
Правильные тесты для работы с участниками поездки

VictorTikhonov committed:
Убран DisplayName, переименованы тесты для ExpenseServiceAddParticipantsTest

VictorTikhonov committed:
Убран DisplayName, переименованы тесты для ExpenseServiceCreateExpenseTest

VictorTikhonov committed:
Убран DisplayName, переименованы тесты для ExpenseServiceDeleteExpenseTest

VictorTikhonov committed:
Убран DisplayName, переименованы тесты для ExpenseServiceDeleteParticipantTest

VictorTikhonov committed:
Убран DisplayName, переименованы тесты для ExpenseServiceGetAllExpensesTest

VictorTikhonov committed:
Убран DisplayName, переименованы тесты для ExpenseServiceUpdateExpenseTest

VictorTikhonov committed:
Убран DisplayName, переименованы тесты для MemberExpenseServiceTest

VictorTikhonov committed:
Небольшие правки

VictorTikhonov committed:
Небольшие правки в ExpenseController

VictorTikhonov committed:
Merge pull request #44 from tbank-ttravels/Expense_tests

Tests

VictorTikhonov committed:
Небольшие правки в TransferServiceTest

EnzoLu1337 committed:
Небольшие правки в TravelMemberServiceTest

VictorTikhonov committed:
Убраны TODO

EnzoLu1337 committed:
Исправлено редактирование юзера

EnzoLu1337 committed:
Merge remote-tracking branch 'origin/deploy' into deploy
